### PR TITLE
Show build errors for all dependency projects

### PR DESCRIPTION
**Pull request by ChengTian**
parent https://github.com/aspnet/dnx/issues/583

When building a project, show diagnostics of dependency projects as well.

Tested with both command-line and VS.

@davidfowl

### CODE REVIEW

**Comment by ChengTian** [comment]
One drawback of doing this is that we will get duplicated error messages in VS error list:
In the example below, `ClassLibrary1` is a dependency of `ConsoleApp6`. The syntax error in `ClassLibrary1` is reported by both `ClassLibrary1` and `ConsoleApp6`

capture

File: src/Microsoft.Framework.PackageManager/Building/BuildContext.cs [hunk 1]
```diff
@@ -54,15 +54,19 @@ public void Initialize(IReport report)
 
         public bool Build(List<ICompilationMessage> diagnostics)
         {
+            var allProjectDiagnostics = DiagnosticsUtils.GetAllReferenceProjectDiagnostics(
+                _applicationHostContext.LibraryManager,
+                _project.Name);
+            diagnostics.AddRange(allProjectDiagnostics);
+            if (diagnostics.HasErrors())
+            {
+                return false;
```
Comment (davidfowl) [inline src/Microsoft.Framework.PackageManager/Building/BuildContext.cs]: Dont' short circuit.

File: src/Microsoft.Framework.Runtime.Compilation.Common/DiagnosticsUtils.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+namespace Microsoft.Framework.Runtime
+{
+    internal static class DiagnosticsUtils
+    {
+        public static List<ICompilationMessage> GetAllReferenceProjectDiagnostics(ILibraryManager libraryManager,
+            string projectName)
+        {
+            var projectExport = libraryManager.GetAllExports(projectName);
+            var metadataProjectRefs = projectExport.MetadataReferences .OfType<IMetadataProjectReference>();
+            return metadataProjectRefs.SelectMany(x => x.GetDiagnostics().Diagnostics).ToList();
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime.Compilation.Common/DiagnosticsUtils.cs]: .Distinct(CreateAComparerForICompilationMessage)

File: src/Microsoft.Framework.Runtime.Compilation.Common/DiagnosticsUtils.cs [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+namespace Microsoft.Framework.Runtime
+{
+    internal static class DiagnosticsUtils
+    {
+        public static List<ICompilationMessage> GetAllReferenceProjectDiagnostics(ILibraryManager libraryManager,
+            string projectName)
+        {
+            var projectExport = libraryManager.GetAllExports(projectName);
+            var metadataProjectRefs = projectExport.MetadataReferences .OfType<IMetadataProjectReference>();
```
Comment (pranavkm) [inline src/Microsoft.Framework.Runtime.Compilation.Common/DiagnosticsUtils.cs]: Formatting.

**Comment by ChengTian** [comment]
@davidfowl , I talked to @PradeepKadubandi and Tooling is ready to react to this change. Please sign off.

The deduplication of errors shared by different frameworks should be resolved in another thread: https://github.com/aspnet/dnx/issues/1710

**Comment by ChengTian** [comment]
@davidfowl , all external CIs passed

**Comment by ChengTian** [comment]
@davidfowl , we talked about the problem of doing `dnu publish --no-source` when we have transitive diagnostics retrieval.

I checked the behavior of msbuild. It builds projects in topological order and exits as soon as it sees a failed project.

Do we want to modify `dnu publish` to do the same thing?

Given Projects A -> B -> C

We build the projects in the order of C B A and if C has errors, we exits immediately and B has no chance to show the duplicated transitive errors from C.

**Comment by davidfowl** [comment]
We're not doing this anymore

**Pull request closed without merging**